src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm

```diff
@@ -264,27 +335,35 @@ - (void)_callNextReadyWorkItem
         while (_items.count >= 2) {
             MTRAsyncWorkItem * nextWorkItem = _items[1];
             if (!nextWorkItem.batchingHandler || nextWorkItem.batchingID != workItem.batchingID) {
-                break; // next item is not eligible to merge with this one
+                goto done; // next item is not eligible to merge with this one
             }
 
-            BOOL fullyMerged = NO;
-            batchingHandler(workItem.batchableData, nextWorkItem.batchableData, &fullyMerged);
-            if (!fullyMerged) {
-                break; // not removing the next item, so we can't merge anything else
+            switch (batchingHandler(workItem.batchableData, nextWorkItem.batchableData)) {
+            case MTRNotBatched:
+                goto done; // can't merge anything else
+            case MTRBatchedPartially:
+                MTR_LOG_DEFAULT("MTRAsyncWorkQueue<%@> partially merged work item %tu into %tu",
+                    context.description, nextWorkItem.uniqueID, workItem.uniqueID);
+                goto done; // can't merge anything else
+            case MTRBatchedFully:
+                MTR_LOG_DEFAULT("MTRAsyncWorkQueue<%@> fully merged work item %tu into %tu",
+                    context.description, nextWorkItem.uniqueID, workItem.uniqueID);
+                [_items removeObjectAtIndex:1];
+                continue; // try to batch the next item (if any)
             }
-
-            [_items removeObjectAtIndex:1];
         }
+    done:;
```

Comment:
Do we need the semicolon here?

Reply:
That's to make it a statement, otherwise I get a compiler warning.